codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -101,89 +79,141 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
      *     {@link ChannelHandler.Sharable}.
      * @param streamGroup event loop for registering child channels
      */
-    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
-            EventLoopGroup streamGroup) {
-        this(server, streamHandler, streamGroup, new DefaultHttp2FrameWriter());
-    }
-
-    Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
-            EventLoopGroup streamGroup, Http2FrameWriter frameWriter) {
+    public Http2MultiplexCodec(boolean server,
+                               EventLoopGroup streamGroup,
+                               ChannelHandler streamHandler) {
         if (!streamHandler.getClass().isAnnotationPresent(Sharable.class)) {
             throw new IllegalArgumentException("streamHandler must be Sharable");
         }
+
+        this.server = server;
         this.streamHandler = streamHandler;
         this.streamGroup = streamGroup;
-        Http2Connection connection = new DefaultHttp2Connection(server);
-        frameWriter = new Http2OutboundFrameLogger(frameWriter, HTTP2_FRAME_LOGGER);
-        Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
-        Http2FrameReader reader = new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), HTTP2_FRAME_LOGGER);
-        Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, reader);
-        decoder.frameListener(new FrameListener());
-        http2Handler = new InternalHttp2ConnectionHandler(decoder, encoder, new Http2Settings());
-        http2Handler.connection().addListener(new ConnectionListener());
-        streamInfoKey = http2Handler.connection().newKey();
     }
 
-    Http2ConnectionHandler connectionHandler() {
-        return http2Handler;
-    }
-
-    /**
-     * Save context and load any dependencies.
-     */
     @Override
     public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
         this.ctx = ctx;
-        ctx.pipeline().addBefore(ctx.executor(), ctx.name(), null, http2Handler);
-        http2HandlerCtx = ctx.pipeline().context(http2Handler);
     }
 
-    /**
-     * Clean up any dependencies.
-     */
     @Override
-    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
-      ctx.pipeline().remove(http2Handler);
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        if (!(cause instanceof StreamException)) {
+            ctx.fireExceptionCaught(cause);
+            return;
+        }
+
+        StreamException streamEx = (StreamException) cause;
+        try {
+            Http2StreamChannel childChannel = childChannels.get(streamEx.streamId());
+            if (childChannel != null) {
+                childChannel.pipeline().fireExceptionCaught(streamEx);
+            }
```

Comment:
should we log the exception at least if we have no childChannel to propagate it too ?